client/transaction-pool/src/lib.rs

```diff
@@ -130,3 +175,158 @@ impl<PoolApi, Block> TransactionPool for BasicPool<PoolApi, Block>
+		block: N,
+		revalidate_time_period: Option<std::time::Duration>,
+		revalidate_block_period: Option<N>,
+	) -> bool {
+		match *self {
+			Self::NotScheduled => {
+				*self = Self::Scheduled(
+					revalidate_time_period.map(|period| Instant::now() + period),
+					revalidate_block_period.map(|period| block + period),
+				);
+				false
+			},
+			Self::Scheduled(revalidate_at_time, revalidate_at_block) => {
+				let is_required = revalidate_at_time.map(|at| Instant::now() >= at).unwrap_or(false)
+					|| revalidate_at_block.map(|at| block >= at).unwrap_or(false);
+				if is_required {
+					*self = Self::InProgress;
+				}
+				is_required
+			},
+			Self::InProgress => false,
+		}
+	}
+}
+
+impl<PoolApi, Block> MaintainedTransactionPool for BasicPool<PoolApi, Block>
+where
+	Block: BlockT,
+	PoolApi: 'static + sc_transaction_graph::ChainApi<Block=Block, Hash=Block::Hash, Error=error::Error>,
+{
+	fn maintain(&self, id: &BlockId<Self::Block>, retracted: &[BlockHash<Self>]) -> Pin<Box<dyn Future<Output=()> + Send>> {
+		let id = id.clone();
+		let pool = self.pool.clone();
+		let api = self.api.clone();
+
+		let block_number = match api.block_id_to_number(&id) {
+			Ok(Some(number)) => number,
+			_ => {
+				log::trace!(target: "txqueue", "Skipping chain event - no numbrer for that block {:?}", id);
+				return Box::pin(ready(()));
+			}
+		};
+
+		let (is_revalidation_required, is_resubmit_required) = {
+			let mut lock = self.revalidation_strategy.lock();
+			(
+				lock.is_required(
+					block_number,
+					Some(std::time::Duration::from_secs(60)),
+					Some(20.into()),
+				),
+				lock.resubmit_required()
+			)
+		};
+
+		let revalidation_status = self.revalidation_strategy.clone();
+		let revalidation_amount = revalidation_status.lock().amount();
+		let retracted = retracted.to_vec();
+
+		async move {
+			let hashes = api.block_body(&id).await
+				.unwrap_or_else(|e| {
+					log::warn!("Prune known transactions: error request {:?}!", e);
+					None
+				})
+				.unwrap_or(Vec::new())
+				.into_iter()
+				.map(|tx| pool.hash_of(&tx))
+				.collect::<Vec<_>>();
+
+			if let Err(e) = pool.prune_known(&id, &hashes) {
+				log::warn!("Cannot prune known in the pool {:?}!", e);
+			}
+
+			if is_resubmit_required {
+				let mut resubmit_transactions = Vec::new();
+
+				for retracted_hash in retracted.into_iter() {
+					let txes = api.block_body(&BlockId::hash(retracted_hash.clone())).await
+						.unwrap_or(None)
+						.unwrap_or(Vec::new());
+					for tx in txes {
+						resubmit_transactions.push(tx)
+					}
+				}
+				if let Err(e) = pool.submit_at(&id, resubmit_transactions, true).await {
+					log::debug!(target: "txpool",
+						"[{:?}] Error re-submitting transactions: {:?}", id, e
+					)
+				}
+			}
+
+			if is_revalidation_required {
+				if let Err(e) = pool.revalidate_ready(&id, revalidation_amount).await {
+					log::warn!("revalidate ready failed {:?}", e);
+				}
+			}
+
+			revalidation_status.lock().clear();
```

Comment:
That can get pretty racy, can't it? If we ever get two `maintain` runs in parallel they will mess up each other statuses.

Reply:
Light strategy revalidation can't run in parallel, since it is single-threaded, but I don't think we can enforce it here

Reply:
I'll add an issue for that.
We might want to run light strategy out of light client context actually